### DISCUSSION TITLE
An L-System example

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#555] [Show and tell]
I have been teaching this in so many classes over the years, and about 4 times this month :smile:. Also, this is perfect for tweaking with the wonderful new [Live Coding feature](https://py5coding.org/content/live_coding.html) :wink: ! 

I really like doing L-Systems with Python + Processing, I always felt like my dict approach to the substitution rules was far better than the rules with arrays in Processing Java examples. 

![out](https://github.com/user-attachments/assets/684d7cab-cff1-4f62-8187-b7d4d9d6c7e1)

```python
import py5

step = 8
axiom = 'F'
rules = {
    'F': 'FF+[+F-F-FO]-[-F+F+FO]',
    }
angle = 22.5
iterations = 4

def setup():
    py5.size(600, 600, py5.P3D)  # canvas 3D
    calculate_sequence()

def calculate_sequence():
    global sequence
    starting_sequence = axiom
    for i in range(iterations):
        sequence = ''
        for symbol in starting_sequence:
            sequence = sequence + rules.get(symbol, symbol)
        starting_sequence = sequence
    print(len(sequence))
    
def draw():
    #global angle
    #angle = py5.mouse_x / 5
    py5.background(240)
    py5.translate(py5.width / 2, py5.height - 50)
    py5.rotate_y(py5.frame_count / 30) # the main rotation on Y
    
    for symbol in sequence:
        if symbol == 'F':
            py5.line(0, 0, 0, -step)
            py5.translate(0, -step)
            py5.rotate_y(py5.radians(angle)) # every line twists
        elif symbol == '+':
            py5.rotate(py5.radians(-angle))
        elif symbol == '-':
            py5.rotate(py5.radians(angle))
        elif symbol == '[':
            py5.push_matrix()
        elif symbol == ']':
            py5.pop_matrix()
        elif symbol == 'O':
            py5.fill(255, 0, 0)
            py5.no_stroke()
            py5.circle(0, 0, step / 2)
            py5.stroke(0)
            
        
py5.run_sketch()
```

To learn more about L-Systems and related stuff:
- https://en.wikipedia.org/wiki/L-system
- Przemysław Prusinkiewicz and Aristid Lindenmayer. [The Algorithmic Beauty of Plants](http://algorithmicbotany.org/papers/#abop)
- http://www.paulbourke.net/fractals/lsys/
- https://www.cgjennings.ca/articles/l-systems/
- [A py5  *imported mode* example]( https://abav.lugaralgum.com/material-aulas/Processing-Python-py5/LSystem.html)
- [Inside chapter 8 of Shiffman's Nature of Code](https://natureofcode.com/fractals/#l-systems)

**Comment by vsquared** [comment]
Nice work.  Impressive demo.

**Comment by binitshah** [comment]
Wow, impressive!

**Comment by hx2A** [comment]
@villares , this is wonderful, and exactly what I was hoping people would use the Live Coding feature for!

**Comment by villares** [comment]
Tomorrow I'm going to present an L-System demo, so, to gear up, today I tried to find a nice variation of these rules with the red circles...

![sketch_2025_04_02](https://github.com/user-attachments/assets/f1007fb3-8c8f-48c4-885f-f06d4a9d43f1)

```python
import py5

step = 10
axiom = 'F'
rules = {
#    'F': 'FF+[+F-F-FO]-[-F+F+FO]',
   'F': 'F[+F-F+FO]F[-F+F-FO]',
    }
angle = 30
iterations = 4

def setup():
    py5.size(600, 600)
    calculate_sequence()

def calculate_sequence():
    global sequence
    starting_sequence = axiom
    for i in range(iterations):
        sequence = ''
        for symbol in starting_sequence:
            sequence = sequence + rules.get(symbol, symbol)
        starting_sequence = sequence
    print(len(sequence))
    
def draw():
    #global angle
    #angle = py5.mouse_x / 5
    py5.background(0)
    py5.stroke(255)
    py5.translate(py5.width / 2, py5.height - 30)
    
    for symbol in sequence:
        if symbol == 'F':
            py5.line(0, 0, 0, -step)
            py5.translate(0, -step)
        elif symbol == '+':
            py5.rotate(py5.radians(-angle))
        elif symbol == '-':
            py5.rotate(py5.radians(angle))
        elif symbol == '[':
            py5.push_matrix()
        elif symbol == ']':
            py5.pop_matrix()
        elif symbol == 'O':
            with py5.push_style():
                py5.fill(255, 0, 0)
                py5.no_stroke()
                py5.circle(0, 0, step / 2)

def key_pressed():
    py5.save_frame('###.png')
            
        
py5.run_sketch()
```

**Reply by hx2A** [reply]
This looks really nice! The red circles are a nice touch.

Did you use live coding to experiment with this?

**Reply by villares** [reply]
Not this time. But I want to get back to using it more often. I have to test the code snapshot feature and compare with stuff I was also doing (triggering the saving of folders with the code and image result of some variations).
